### DISCUSSION TITLE
Include lwIP its hidden MQTT protocol app

### DISCUSSION THREAD

**Discussion by keever50** [earlephilhower/arduino-pico#2117] [Ideas]
There are many IoT projects that can be made using Arduino-pico, where I made some of them. Most IoT devices use the MQTT protocol to communicate. I believe most users download popular Arduino MQTT libraries, such as pub-sub. While we already have an unused MQTT library built in!

Arduino-pico includes the raspberry pi pico SDK. The pico SDK includes a library called lwIP, which is a lightweight TCP/IP stack. It handles all the networking.
This lwIP library contains ready to use applications where one of the applications is the MQTT protocol.
src/include/lwip/apps/mqtt.h at https://github.com/lwip-tcpip/lwip 

Now, I am not sure how to include this library to Arduino-pico, but I have got it working on the bare pico-SDK. But it is undoubtedly a great addition to Arduino-pico. Ofcourse all the other apps are very interesting as well.

Are there any reasons why this is not included already? Technical limitations? Maybe we didn’t know before?
How could it be implemented? Perhaps the way Arduino handles wifi?

![image](https://github.com/earlephilhower/arduino-pico/assets/9679272/a41d99de-3883-4c15-8b15-059d51651752)

**Comment by earlephilhower** [comment]
It's not built as part of the core, and it doesn't use the standard Arduino interfaces.  While you could use it, I suppose, by adding it to the build and manually using the raw LWIP interfaces, I don't see any reason or benefit to doing so.  The existing Arduino MQTT libraries all seem to work well for users, and have many examples and real-world deployments.

One other reason not to do it would be that the LWIP version only works on raw connections.  You can't use TLS or client certificates like you can with the Arduino-based libraries.  So that's a big negative there.

**Reply by keever50** [reply]
I understand that the arduino libraries are proven. I agree.

But it turns out that MQTT in lwIP 2.1.0 does actually support TLS, according https://www.nongnu.org/lwip/2_1_x/index.html 🤔 
Thought maybe thats a good thing to know.

However, the compatibility is indeed a hell now that we work with arduino stuff.

Oh well, if anyone needs applications like these, the bare SDK might be a better option over arduino anyway.

**Reply by earlephilhower** [reply]
Looks like `altcp` is an `mbedTLS` wrapper.  We don't support that here (run BearSSL since that's what I ported to the ESP8266 way back when), but if you really want to go down this path then check out the official Arduino code.  It runs `mbedTLS` AIUI, so you'd not have to get that ported as well.  Whether the MQTT app runs w/LWIP in `NO_SYS==1` mode, you'll find out. :)  Good luck!

**Reply by keever50** [reply]
Interesting, I had no idea!
Very good to know
